freqtrade/persistence/models.py

```diff
@@ -505,56 +630,135 @@ def recalc_open_trade_value(self) -> None:
         """
         self.open_trade_value = self._calc_open_trade_value()
 
+    def calculate_interest(self, interest_rate: Optional[float] = None) -> Decimal:
+        """
+        : param interest_rate: interest_charge for borrowing this coin(optional).
+        If interest_rate is not set self.interest_rate will be used
+        """
+        # TODO-mg: Need to set other conditions because sometimes self.open_date is not defined, but why would it ever not be set
+        zero = Decimal(0.0)
+        if not (self.borrowed):
+            return zero
+
+        open_date = self.open_date.replace(tzinfo=None)
+        now = datetime.utcnow()
+        # sec_per_day = Decimal(86400)
+        sec_per_hour = Decimal(3600)
+        total_seconds = Decimal((now - open_date).total_seconds())
+        # days = total_seconds/sec_per_day or zero
+        hours = total_seconds/sec_per_hour or zero
+
+        rate = Decimal(interest_rate or self.interest_rate)
+        borrowed = Decimal(self.borrowed)
+        one = Decimal(1.0)
+        twenty_four = Decimal(24.0)
+        four = Decimal(4.0)
+
+        if self.exchange == 'binance':
+            # Rate is per day but accrued hourly or something
+            # binance: https://www.binance.com/en-AU/support/faq/360030157812
+            return borrowed * rate * max(hours, one)/twenty_four  # TODO-mg: Is hours rounded?
+        elif self.exchange == 'kraken':
+            # https://support.kraken.com/hc/en-us/articles/206161568-What-are-the-fees-for-margin-trading-
+            opening_fee = borrowed * rate
+            roll_over_fee = borrowed * rate * max(0, (hours-four)/four)
+            return opening_fee + roll_over_fee
+        elif self.exchange == 'binance_usdm_futures':
+            # ! TODO-mg: This is incorrect, I didn't look it up
+            return borrowed * (rate/twenty_four) * max(hours, one)
+        elif self.exchange == 'binance_coinm_futures':
+            # ! TODO-mg: This is incorrect, I didn't look it up
+            return borrowed * (rate/twenty_four) * max(hours, one)
+        else:
```

Comment:
This is a no-go (adding exchange-specific stuff to the models file) - as it can break too easily.

If there's different fee models - create an enum, save the "fee model" in the database upon trade creation. That'll allow for easier reusability.

Reply:
Are you opposed to having an exchange object instantiated on the `LocalTrade` object and then getting  the interest  by calling something like `exchange.calculateInterest()`? It would be useful for calculating the `leverage` and `borrowed` base on the exchange also

I could also create 2 enums, what folder would these go in?

Reply:
Currently - you check for `self.exchange == 'whatever'` - instead of doing this - check for `self.interrest_mode = InterrestModes.Daily` (or whatever) ... 

that way we can define in the exchange class the calculation method to use - without having to have a seperate case for each exchange.
adding support for new exchanges should NOT require changes to the database model.

Reply:
There is no exchange object that exists on the `LocalTrade` or `Trade` class, all we have to refer to the exchange is the string. Would we create an exchange object in order to do this?

Reply:
no, we assign the enum to a column in the trade / Local trade object - which determines the logic to calculate Interrests.

The moment we create a trade - we will know the exchange (that's done from either backtesting or freqtradebot - both of which have an exchange object) - and that's the moment we assign this (this is something that cannot change).

Reply:
It's a little more complicated that daily/hourly. Binance has a daily interest rate, but the minimum is one hour of interest, whereas kraken has a 4 hour interest rate with a minimum of 4 hours interest

Reply:
What about a delegate?

Reply:
Trade objects are stored (and loaded) from the database - they're not "there all the time" - you can't persist a function (well technically you can - but we will not pickle functions into the database!) - so you'd have to reattach the function each time you load it from the database - which is a quite dangerous task